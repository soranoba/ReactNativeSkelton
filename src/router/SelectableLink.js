/**
 * @author Hinagiku Soranoba <soranoba@gmail.com>
 * @license MIT
 */
import React from 'react';
import { Route, Link } from 'react-router-native';
import PropTypes from 'prop-types';

/**
 * CustomLink of 'react-router', which set `selected` to child.props.
 *
 * - This Component allow only single child.
 * - For currently displayed pages, the props.selected of child will equal to true.
 *
 * See: https://reacttraining.com/react-router/native/example/custom-link
 */
export default class SelectableLink extends React.Component {
  static propTypes = {
    ...Link.propTypes,
    children: PropTypes.node.isRequired,
  };

  render() {
    const { to, children, activeOnlyWhenExact, ...rest } = this.props;
    const link = ({ match }) => {
      return (
        <Link style={{ flex: 1 }} to={to} {...rest}>
          {React.cloneElement(React.Children.only(children), { selected: match && match.isExact })}
        </Link>
      );
    };
    return <Route path={to} exact={activeOnlyWhenExact} children={link} />;
  }
}
