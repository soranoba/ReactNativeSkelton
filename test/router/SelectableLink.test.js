import React from 'react';
import renderer from 'react-test-renderer';
import SelectableLink from '/router/SelectableLink';
import { View } from 'react-native';
import { StaticRouter } from 'react-router';

describe('SelectableLink', () => {
  function testNode(currentPath, linkProps, viewProps) {
    return (
      <StaticRouter location={currentPath} context={{}}>
        <SelectableLink {...linkProps}>
          <View {...viewProps} />
        </SelectableLink>
      </StaticRouter>
    );
  }

  it('set true to props.selected, when link path is currentPath', () => {
    const viewProps = { style: { backgroundColor: 'red' } };
    const rendered = renderer
      .create(testNode('/target', { to: '/target', replace: true }, viewProps))
      .toJSON();
    expect(rendered.children[0].props).toMatchObject(Object.assign(viewProps, { selected: true }));
  });

  it('set false to props.selected, when link path is not currentPath', () => {
    const viewProps = { style: { backgroundColor: 'red' } };
    const rendered = renderer
      .create(testNode('/current', { to: '/target', replace: true }, viewProps))
      .toJSON();
    expect(rendered.children[0].props).toMatchObject(Object.assign(viewProps, { selected: false }));
  });

  it('renders currectly', () => {
    const rendered = renderer.create(testNode('/target', { to: '/target' }, {})).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});
