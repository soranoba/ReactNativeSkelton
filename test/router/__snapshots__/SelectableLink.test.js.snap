// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectableLink renders currectly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    selected={true}
    style={null}
  />
</View>
`;
